packages/gatsby/src/internal-plugins/query-runner/query-watcher.js

```diff
@@ -16,18 +16,24 @@ const { boundActionCreators } = require(`../../redux/actions`)
 const queryCompiler = require(`./query-compiler`).default
 const queue = require(`./query-queue`)
 const normalize = require(`normalize-path`)
+const report = require(`gatsby-cli/lib/reporter`)
 
 exports.extractQueries = () => {
   const state = store.getState()
   const pagesAndLayouts = [...state.pages, ...state.layouts]
-  const components = _.uniq(pagesAndLayouts.map(p => p.component))
+  const components = _.uniq(pagesAndLayouts.map(p => normalize(p.component)))
   const queryCompilerPromise = queryCompiler().then(queries => {
-    components.forEach(component => {
-      const query = queries.get(normalize(component))
-      boundActionCreators.replaceComponentQuery({
-        query: query && query.text,
-        componentPath: component,
-      })
+    queries.forEach((query, component) => {
+      if (_.includes(components, component)) {
+        boundActionCreators.replaceComponentQuery({
+          query: query && query.text,
+          componentPath: component,
+        })
+      } else {
+        report.warn(
+          `Query in component "${component}" will not be executed. This component is not page or layout.`
```

Comment:
+1 on the warning, maybe we can make it a bit more exhaustive.

> GraphQL query in component "${component}" will not be run! Queries are only executed for Page or Layout components. Instead of a query, co-locate a GraphQL fragment and compose that fragment into the query (or other fragment) of the top-level page or layout that renders this component. For more info on fragments and composition see: http://graphql.org/learn/queries/#fragments

Reply:
TODO: we really need gatsby specific docs on this

Reply:
Yeah definitely this needs better message (noone reads PR descriptions - I'm guilty of it as well :D ). Will leave it as is for now so your comment won't be hidden in "outdated comment" thingie and will change it later.

Reply:
@pieh are you suggesting we merge this then add @jquense's message in a follow-up PR? Why not just change it in this PR? Also 👍  to documenting fragments directly. @m-allanson has been making strides on some work related to graphql docs.

Reply:
I will update PR later today, just wanted it to breathe few days for folks to join conversation about warning message and didn't want to update PR multiple times.

Reply:
Ah cool :-) 👍